src/lib/gadgets/basic.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Basic gadgets that only use generic gates
+ */
+import type { Field, VarField } from '../field.js';
+import { existsOne, toVar } from './common.js';
+import { Gates } from '../gates.js';
+import { TupleN } from '../util/types.js';
+
+export { assertOneOf };
+
+// TODO: create constant versions of these and expose on Gadgets
+
+/**
+ * Assert that a value equals one of a finite list of constants:
+ * `(x - c1)*(x - c2)*...*(x - cn) === 0`
+ *
+ * TODO: what prevents us from getting the same efficiency with snarky DSL code?
+ */
+function assertOneOf(x: Field, allowed: [bigint, bigint, ...bigint[]]) {
+  let xv = toVar(x);
+  let [c1, c2, ...c] = allowed;
+  let n = c.length;
+  if (n === 0) {
+    // (x - c1)*(x - c2) === 0
+    assertBilinear(xv, xv, [1n, -(c1 + c2), 0n, c1 * c2]);
+    return;
+  }
+  // z = (x - c1)*(x - c2)
+  let z = bilinear(xv, xv, [1n, -(c1 + c2), 0n, c1 * c2]);
+
+  for (let i = 0; i < n; i++) {
```

Comment:
Because this only uses generics, I guess the maximum length of the list doesn't really matter. Meaning, you don't need to set a bound to X terms so that the underneath constraints are not higher than degree X. In other words, this is just concatenating generics and copy constraints between outputs and inputs among them, right? Whether or not the check becomes too costly in terms of rows, just depends on the zkapp developer criterion, no?

Reply:
yes exactly, just concatenates generic gates